indexing-service/src/main/java/io/druid/indexing/overlord/http/OverlordResource.java

```diff
@@ -629,6 +648,207 @@ public Response getCompleteTasks(
+        workItem.getTaskId(),
+        workItem.getTaskType(),
+        workItem.getCreatedTime(),
+        workItem.getQueueInsertionTime(),
+        workItem.getTaskState(),
+        null,
+        workItem.getLocation(),
+        workItem.getDataSource(),
+        null
+    );
+    List<TaskStatusPlus> transformedActiveList = Lists.transform(activeTasks, transformFunc);
+    final List<TaskStatusPlus> allTasks = Lists.newArrayList();
+    allTasks.addAll(transformedActiveList);
+    allTasks.addAll(completedTasks);
+
+    Collection<? extends TaskStatusPlus> authorizedList = securedTaskStatusPlus(
+        allTasks,
+        dataSource,
+        req
+    );
+    return Response.ok(authorizedList).build();
+  }
+
+  private List<TaskStatusPlus> getCompletedTasks(
+      String interval,
+      Integer maxCompletedTasks,
+      String dataSource,
+      HttpServletRequest req
+  )
+  {
+    Duration duration = null;
+    if (interval != null) {
+      final Interval theInterval = Intervals.of(interval.replace("_", "/"));
+      duration = theInterval.toDuration();
+    }
+
+    final Function<String, Task> taskFunction = taskId -> {
+      final Optional<Task> optionalTask = taskStorageQueryAdapter.getTask(taskId);
+      if (!optionalTask.isPresent()) {
+        throw new WebApplicationException(
+            Response.serverError().entity(
+                StringUtils.format("No task information found for task with id: [%s]", taskId)
+            ).build()
+        );
+      }
+      return optionalTask.get();
+    };
+
+    final List<TaskStatus> recentlyFinishedTasks = taskStorageQueryAdapter.getRecentlyFinishedTaskStatuses(
+            maxCompletedTasks, duration);
+
+    final List<TaskStatusPlus> completeTasks = Lists.newArrayList(Iterables.transform(
+        recentlyFinishedTasks,
+        status -> {
+          final Pair<DateTime, String> pair = taskStorageQueryAdapter.getCreatedDateAndDataSource(status.getId());
+          return new TaskStatusPlus(
+              status.getId(),
+              taskFunction.apply(status.getId()).getType(),
+              pair.lhs,
+              // Would be nice to include the real queue insertion time, but the
+              // TaskStorage API doesn't yet allow it.
+              DateTimes.EPOCH,
+              status.getStatusCode(),
+              status.getDuration(),
+              TaskLocation.unknown(),
+              pair.rhs,
+              status.getErrorMsg()
+          );
+        }
+    ));
+
+    return completeTasks;
+  }
+
+  private List<TaskRunnerWorkItem> getAllActiveTasks(HttpServletRequest req)
+  {
+    final List<Task> allActiveTasks = taskStorageQueryAdapter.getActiveTasks();
+    final List<TaskRunnerWorkItem> allTasks = Lists.newArrayList();
+    for (final Task task : allActiveTasks) {
+      allTasks.add(new AnyTaskBuilder(
+          task.getId(),
+          task.getType(),
+          SettableFuture.create(),
+          task.getDataSource(),
+          TaskState.PENDING
```

Comment:
Why is the state fixed to pending?

Reply:
Initially, the state is set to PENDING as a placeholder for all active tasks, later in the method, we filter out the RUNNING and WAITING tasks from this list, so the remaining tasks would be actually in PENDING state.